rsfg/src/lexer.rs

```diff
@@ -6,175 +6,235 @@ static SPACE: &str = " \t";
 
 // Rules: A-Z,a-z
 fn is_id_1st(c: char) -> bool {
-	c >= 'A' && c <= 'z'
+    c >= 'A' && c <= 'z'
 }
+
 // A-Z or 0-9
 fn is_id(c: char) -> bool {
-	is_id_1st(c) || (c >= '0' && c <= '9')
+    is_id_1st(c) || (c >= '0' && c <= '9')
+}
+
+enum NextSymbolType {
+    None,
+    SymbolOrId(char),
+    Space(char),
+    Digit(char),
+    Slash,
+    Newline,
+    LParen,
+    RParen,
+    Colon,
+    Comma,
+    Quote,
+    Unknown(char),
+}
+
+#[derive(Debug)]
+struct Lexer<'src> {
+    source: &'src str,
+    tokens: Vec<Token>,
+    spaces_count: usize,
+    rchars: Vec<char>,
+}
+
+impl<'src> Lexer<'src> {
+    pub fn new(source: &'src str) -> Lexer<'src> {
+        Lexer {
+            source,
+            tokens: Vec::new(),
+            spaces_count: 0,
+            rchars: source.chars().rev().collect(),
+        }
+    }
+
+    fn is_eof(&self) -> bool {
+        self.rchars.len() == 0
+    }
+
+    fn next_symbol_type(&mut self) -> NextSymbolType {
+        use NextSymbolType::*;
+
+        if self.is_eof() {
+            return None;
+        }
+
+        let c = self.rchars.pop().expect("not supposed to be at EOF");
```

Comment:
```rust
let c = match self.rchars.pop() {
    Some(c) => c,
    None => return None, // or EOF
};
```
agrees with how i do everything else, and then u can drop the `is_eof` function

Reply:
Yes, definitely nicer. Done.